promise.go

```diff
@@ -85,3 +85,41 @@ func (p *Promise) Result() *Value {
 	val := &Value{ptr, p.ctx}
 	return val
 }
+
+// Then accepts 1 or 2 callbacks.
+// The first is invoked when the promise has been fulfilled.
+// The second is invoked when the promise has been rejected.
+// The returned Promise resolves after the callback finishes execution.
+//
+// V8 only invokes the callback when processing "microtasks".
+// The default MicrotaskPolicy processes them when the call depth decreases to 0.
+// Call (*Context).PerformMicrotaskCheckpoint to trigger it manually.
+func (p *Promise) Then(cbs ...FunctionCallback) *Promise {
+	p.ctx.register()
+	defer p.ctx.deregister()
+
+	var ptr C.ValuePtr
+	switch len(cbs) {
+	case 1:
+		cbID := p.ctx.iso.registerCallback(cbs[0])
+		ptr = C.PromiseThen(p.ptr, C.int(cbID))
+	case 2:
+		cbID1 := p.ctx.iso.registerCallback(cbs[0])
+		cbID2 := p.ctx.iso.registerCallback(cbs[1])
+		ptr = C.PromiseThen2(p.ptr, C.int(cbID1), C.int(cbID2))
+
+	default:
+		panic("1 or 2 callbacks required")
```

Comment:
Would like to avoid `panic` within this library and also just ignore Callbacks if the length in > 2. Can we just `return p` and make it a noop?

Reply:
OK. I think we really don't see eye to eye on this design question, since I feel adverse to this. IMO this goes against the typical design ethos & common practice.. that is, ignoring unexpected input and carrying on. But it's your library, and in this case the chance of mistakes by the caller seems low. Updated

Reply:
That's a fair comment and I agree. Would it be better to return an `error`? 🤔 

The downside is you loose the ability to do method chaining (which is not generally idiomatic Go)

I'm just trying to avoid using `panic`

I'm open to other's input? @tmc @cleiner @zwang

Reply:
I prefer not panic solution as this is a library to be used by other app. Good not to panic.

Reply:
I think it would be better to panic, hence the code :)

I would refer you again to examples in the standard library:
https://pkg.go.dev/strings#Repeat

There are no fewer than 3 functions in the `strings` package that are documented to panic when given an invalid input. It would be frustrating for callers if they had to handle an error return unnecessarily. It's ok for the standard library to panic but not for `v8go`? The argument that panics are not allowed in public interfaces at all must not be telling the whole story.

Searching for more information on the topic, I agree with the point of view presented here by Axel:
https://groups.google.com/g/golang-nuts/c/5S324hgLdfI/m/XpWmVQGJCAAJ

Personally I've been using Go since 1.0 (... 10 years now?) and I've seen a lot of libraries. I do not believe the nil checks and error returns  used in v8go are common practice at all. Can you point to popular packages that do this?

Reply:
I'm leaning to your point of view @robfig.

I find this example in the stdlib interesting: https://golang.org/src/io/io.go?s=13817:13895#L382
They could have returned an error as the function returns an error, but they still choose to panic.

For Isolate, and Context I was thinking (at the time of first writing) that I would add error handling for those constructors "at some point" and did not want to introduce breaking API changes. I've seen this in the stdlib before, just can't remember where.

Removing those error returns from those constructors would make caller code a lot simpler.

What I like about Axel's comment in that thread is there is a clear distinction between a panic and an error.

Reply:
I also think it's better to make bugs in the calling code obvious and panic than to return an error (or even ignore them).

Reply:
Ok; I think we are all in agreement now. Sorry for the back-and-forth @robfig can you revert your last commit? ie. place back the panic.

Reply:
Sure, happy to do it. Updated.

Last note - a lot of the errors returned in v8go are as a result of nil checks. In the standard library, it doesn't even perform nil checks, relying on the code to panic on nil as a side effect. 

For example, it looks to me like `io.WriteString(w Writer, s string)` will panic if either of the arguments are nil. 
https://golang.org/src/io/io.go?s=13817:13895#L307

I know there is a practice (which I've seen in Java in particular) to aggressively check arguments for nil on entry to every function. I haven't really seen that done in Go, and I can't say that I've missed it.

Reply:
So it seem the conclusion is: panic is OK when arguments to the API are invalid, but for results of running JavaScript, there will still be `err` returned? So if JavaScript code throws a top-level exception, there will still be `err` and not panic?

Reply:
Yea, that seems like the right approach; JS errors would still return error

Reply:
OK. So will you then change the API now? Like `template.Set` currently returns `err`. It seems it should not anymore and just panic?